python/pyarrow/_substrait.pyx

```diff
@@ -15,35 +15,183 @@
 # specific language governing permissions and limitations
 # under the License.
 
+import base64
+import cloudpickle
+import inspect
+
 # cython: language_level = 3
-from cython.operator cimport dereference as deref
+from cython.operator cimport dereference as deref, preincrement as inc
 
+from pyarrow import compute as pc
 from pyarrow import Buffer
+from pyarrow.lib import frombytes, tobytes
 from pyarrow.lib cimport *
 from pyarrow.includes.libarrow cimport *
 from pyarrow.includes.libarrow_substrait cimport *
+from pyarrow._compute cimport FunctionRegistry
+
+
+from pyarrow._exec_plan cimport is_supported_execplan_output_type, execplan
+from pyarrow._compute import make_function_registry
+
+
+def make_extension_id_registry():
+    cdef:
+        shared_ptr[CExtensionIdRegistry] c_extid_registry
+        ExtensionIdRegistry registry
+
+    with nogil:
+        c_extid_registry = MakeExtensionIdRegistry()
+
+    return pyarrow_wrap_extension_id_registry(c_extid_registry)
+
+
+def _get_udf_code(func):
+    return frombytes(base64.b64encode(cloudpickle.dumps(func)))
```

Comment:
You can instead cimport `pickle` from `pyarrow.compat`, which will use cloudpickle if available, otherwise regular pickle.